src/globus_sdk/services/auth/consents.py

```diff
@@ -0,0 +1,311 @@
+        requirements, as defined by a scope object are satisfied.
+
+    Consents should be retrieved from the AuthClient's `get_consents` method.
+
+    Example usage:
+    >>> auth_client = AuthClient(...)
+    >>> identity_id = ...
+    >>> forest = auth_client.get_consents(identity_id).to_forest()
+    >>>
+    >>> # Check whether the forest contains a scope relationship
+    >>> dependent_scope = GCSCollectionScopeBuilder(collection_id).data_access
+    >>> scope = f"{TransferScopes.all}[{dependent_scope}]"
+    >>> forest.contains_scopes(scope)
+
+
+    The following diagram demonstrates a Consent Forest in which a user has consented
+        to a client ("CLI") initiating transfers against two collections, both of which
+        require a "data_access" dynamic scope.
+    Contained Scope String:
+        `transfer:all[<collection1>:data_access <collection2>:data_access]`
+    ```
+    [Consent A          ]    [Consent B                       ]
+    [Client: CLI        ] -> [Client: Transfer                ]
+    [Scope: transfer:all]    [Scope: <collection1>:data_access]
+            |
+            |                [Consent C                       ]
+            |--------------> [Client: Transfer                ]
+                             [Scope: <collection2>:data_access]
+    ```
+    """
+
+    def __init__(self, consents: t.Iterable[t.Mapping[str, t.Any] | Consent]):
+        """
+        :param consents: An iterable of consent data objects. Typically, this will be
+            a ConsentForestResponse retrieved via `auth_client.get_consents(identity)`.
+            This iterable may contain either raw consent data as a dict or pre-loaded
+            Consents.
+        """
+        self.nodes = [
+            consent if isinstance(consent, Consent) else Consent.load(consent)
+            for consent in consents
+        ]
+        # Build an index on consent id for constant time lookups
+        self._node_by_id = {node.id: node for node in self.nodes}
+
+        self.edges = self._compute_edges()
+        self.trees = self._build_trees()
+
+    def _compute_edges(self) -> dict[int, set[int]]:
+        """
+        Compute the edges of the forest mapping parent -> child.
+
+        A consent's parent node id is defined as the penultimate element of the
+            consent's dependency path.
+        If the dependency list is of length 1, the consent is a has no parent.
+        """
+        edges: dict[int, set[int]] = {node.id: set() for node in self.nodes}
+        for node in self.nodes:
+            if len(node.dependency_path) > 1:
+                parent_id = node.dependency_path[-2]
+                try:
+                    edges[parent_id].add(node.id)
+                except KeyError as e:
+                    raise ConsentParseError(
+                        f"Failed to compute forest edges. Missing parent node: {e}. "
+                        f"Consents: {self.nodes}."
```

Comment:
Are you proposing the following?

```python
class ConsentParseError(Exception):
  def __init__(self, message: str, consents: list[Consent]):
    super().__init__(message)
    self.consents = consents

...

    except KeyError as e:
      raise ConsentParseError(
        f"Failed to compute forest edges. Missing parent node: {e}. ",
        consents=self.nodes
      )
```

Reply:
Yeah, pretty much exactly that. I think that might be what I prefer in the other case as well.

My concerns are:
- overwhelming a developer with too much information at once
- doing a potentially expensive string formatting operation during error handling

Reply:
I think it'll be pretty useful to keep some representation of the consents that we failed to parse in the message itself.

I'm trying to think proactively in these error messages about tickets. The user will most likely just copy/paste whatever this error serializes to into the ticket body so if we move it to this proposed attribute model, the info we'll get in a ticket is:

> "Failed to compute forest edges. Missing parent node: KeyError 12345"

which doesn't feel like enough info to actually help a user.

---

* It is a lot of info but it's a lot info at the tail end of formatted string; The information that "this failed to compute forest edges" should be front and center even if there is a copy of the data we failed to parse at the end.
* It could be potentially expensive that's a fair point. The ticket case above is most of what gives me pause about updating this.

Reply:
I don't want to discount the usage mode you're describing -- we definitely will see user tickets if this fails. However, I think that's allowing a concern about how we manage user support to influence the design of the SDK as a programmer's toolbox in a potentially negative way.

An error object which offers a short summary string and has structured information in additional fields lets the programmer writing an `except` block decide which information to show and how to inspect or dispatch on it. I'm wary of putting a big info dump into the `__repr__` and therefore traceback for an error, but I think the question of whether or not it's "good design" revolves more around how well the exception object itself exposes structured data for inspection.

This is something I've seen as a design flaw in other tools, in which the programmer is forced to do error parsing over some string because the error characteristics are only expressed inside of a large, composite string. The issue can be resolved to a great degree by making the error message contain information which is decomposed into other fields on the error object. i.e. `error.message = f"{error.reason} while parsing {error.data}"` or something like that.

If we capture `nodes=nodes`, `reason="Failed to compute forest edges. Missing parent node."`, etc. I think my main concern here about destructuring structured data will be allayed.

(These names, fields, and format are suggestions, feel free to choose others. :slightly_smiling_face: )

---

I also admit that I'm suspicious that it will be beneficial to have that large blob of data at the end of the string. I think a very likely outcome is that we see error reports containing
```
Failed to compute forest edges. Missing parent node: KeyError 12345. Consents: {11111:
```
_truncated_ so that we haven't really gained the full info.

Reply:
Moved data out of the message.